test/e2e_node/image_pull_test.go

```diff
@@ -0,0 +1,286 @@
+			var httpdPulled, httpdNewPulled pulledStruct
+			for _, event := range events.Items {
+				var err error
+				if event.Reason == kubeletevents.PulledImage {
+					if event.InvolvedObject.Name == testpod.Name {
+						httpdPulled, err = getDurationsFromPulledEventMsg(event.Message)
+						framework.ExpectNoError(err)
+					} else if event.InvolvedObject.Name == testpod2.Name {
+						httpdNewPulled, err = getDurationsFromPulledEventMsg(event.Message)
+						framework.ExpectNoError(err)
+					}
+				}
+			}
+
+			// as this is parallel image pulling, the waiting duration should be similar with the pulled duration.
+			// use 1.2 as a common ratio
+			if float32(httpdNewPulled.pulledIncludeWaitingDuration/time.Millisecond)/float32(httpdNewPulled.pulledDuration/time.Millisecond) > 1.2 {
+				framework.Failf("the pull duration including waiting %v should be similar with the pulled duration %v",
+					httpdNewPulled.pulledIncludeWaitingDuration, httpdNewPulled.pulledDuration)
+			}
+			if float32(httpdPulled.pulledIncludeWaitingDuration/time.Millisecond)/float32(httpdPulled.pulledDuration/time.Millisecond) > 1.2 {
+				framework.Failf("the pull duration including waiting %v should be similar with the pulled duration %v",
+					httpdPulled.pulledIncludeWaitingDuration, httpdPulled.pulledDuration)
+			}
+		})
+
+	})
+})
+
+var _ = SIGDescribe("Pull Image", framework.WithSerial(), nodefeature.MaxParallelImagePull, func() {
+
+	f := framework.NewDefaultFramework("serialize-pull-image-test")
+	f.NamespacePodSecurityLevel = admissionapi.LevelPrivileged
+
+	ginkgo.Context("serialize image pull", func() {
+		// this is the default behavior now.
+		tempSetCurrentKubeletConfig(f, func(ctx context.Context, initialConfig *kubeletconfig.KubeletConfiguration) {
+			initialConfig.SerializeImagePulls = true
+			initialConfig.MaxParallelImagePulls = ptr.To[int32](1)
+		})
+
+		httpdImage := imageutils.GetE2EImage(imageutils.Httpd)
+		httpdNewImage := imageutils.GetE2EImage(imageutils.HttpdNew)
+		var testpod, testpod2 *v1.Pod
+
+		ginkgo.BeforeEach(func(ctx context.Context) {
+			testpod = &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "httpd",
+					Namespace: f.Namespace.Name,
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{{
+						Name:            "httpd",
+						Image:           httpdImage,
+						ImagePullPolicy: v1.PullAlways,
+					}},
+					RestartPolicy: v1.RestartPolicyNever,
+				},
+			}
+			testpod2 = &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "httpd2",
+					Namespace: f.Namespace.Name,
+				},
+				Spec: v1.PodSpec{
+					Containers: []v1.Container{{
+						Name:            "httpd-new",
+						Image:           httpdNewImage,
+						ImagePullPolicy: v1.PullAlways,
+					}},
+					RestartPolicy: v1.RestartPolicyNever,
+				},
+			}
+
+			ginkgo.By("cleanup images")
+			_ = RemoveImage(testpod.Spec.Containers[0].Image)
+			_ = RemoveImage(testpod2.Spec.Containers[0].Image)
+		})
```

Comment:
This looks similar to the code from the previous test. Can we use it as a function?

Reply:
Good idea.